src/main/java/org/mifos/connector/mpesa/camel/routes/PaybillRoute.java

```diff
@@ -27,80 +29,82 @@
                 .setBody(e -> {
                     PaybillRequestDTO paybillRequestDTO = e.getIn().getBody(PaybillRequestDTO.class);
                     //Getting the ams name
-                    String amsName = e.getIn().getHeader("amsName", String.class);
-                    String dfspId = e.getIn().getHeader("dfspId", String.class);
-                    String currency = e.getIn().getHeader("currency", String.class);
-                    //Channel URL
-                    logger.debug("Channel URL : {}", channelUrl);
+                    String businessShortCode = paybillRequestDTO.getShortCode();
+                    String amsName = mpesaPaybillProp.getAMSFromShortCode(businessShortCode);
+                    String amsUrl = mpesaUtils.getAMSUrl(amsName);
+
+                    e.getIn().setHeader("amsUrl", amsUrl);
+                    e.getIn().setHeader("amsName", amsName);
+                    e.getIn().setHeader("accountHoldingInstitutionId", accountHoldingInstitutionId);
                     e.setProperty("channelUrl", channelUrl);
                     e.setProperty("secondaryIdentifier", secondaryIdentifierName);
                     e.setProperty("secondaryIdentifierValue", paybillRequestDTO.getMsisdn());
-                    e.setProperty("dfspId", dfspId);
                     ChannelRequestDTO obj = MpesaUtils.convertPaybillPayloadToChannelPayload(paybillRequestDTO, amsName, currency);
-                    return obj.toString();
+                    String channelRequestDTO = null;
+                    try {
+                        channelRequestDTO = objectMapper.writeValueAsString(obj);
+                    } catch (JsonProcessingException ex) {
+                        throw new RuntimeException(ex);
+                    }
+                    return channelRequestDTO;
                 })
                 .log("MPESA Request Body : ${body}")
                 .toD("${header.channelUrl}" + "/accounts/validate/${header.secondaryIdentifier}/${header.secondaryIdentifierValue}" + "?bridgeEndpoint=true&throwExceptionOnFailure=false")
-                .choice()
-                .when(header("CamelHttpResponseCode").isEqualTo("200"))
                 .log(LoggingLevel.INFO, "Request sent to channel")
+                .setBody(e -> {
+                    String paybillResponseBodyString = e.getIn().getBody(String.class);
+                    JSONObject paybillResponse = new JSONObject(paybillResponseBodyString);
+                    logger.debug("Paybill Response Body : {}", paybillResponseBodyString);
+                    logger.debug("Reconciled : {}", paybillResponse.getBoolean("reconciled"));
+                    GsmaTransfer gsmaTransfer = mpesaUtils.createGsmaTransferDTO(paybillResponse);
+                    e.getIn().setHeader("accountHoldingInstitutionId", paybillResponse.getString("accountHoldingInstitutionId"));
+                    e.getIn().setHeader("amsName", paybillResponse.getString("amsName"));
```

Comment:
Are we still getting the AmS name header? We are planning to infer this from the business short code , arent we?

Reply:
This is after validation success for a given request.